### ISSUE TITLE
feed.rb:237: warning: uncaught exception from callback

### ISSUE THREAD

**Issue by thefonso** [feedjira/feedjira#272]
I'm getting this error...When calling the Blog.update_feed! method inside the Blog.rb model...

/Users/xxxxx/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/feedjira-1.3.0/lib/feedjira/feed.rb:237: warning: uncaught exception from callback

I was following this tutorial...https://binarapps.com/blog/feedjira-with-rails/

and I have a copy of the tutorial app here...https://github.com/thefonso/feedjira-with-rails

can you assist?

**Comment by amatriain** [comment]
What Feedjira method are you invoking where the error is raised? ```Feedjira::Feed.fetch_and_parse```? ```Feedjira::Feed.parse```?

If you're fetching a feed, what is its url? If you're parsing a local xml file, can we take a look?

**Comment by thefonso** [comment]
I'm invoking this method ...https://github.com/thefonso/feedjira-with-rails/blob/master/app/models/blog.rb

```
  def update_feed!
    @newest_entry = entries.order(published: :desc).first
    download_success_date = lambda { |url, feed|
      next unless !@newest_entry || entry.published > @newest_entry.published
      add_new_entry_from_feed(entry)
    }
    Feedjira::Feed.fetch_and_parse(
      url,
      on_success: download_success_date,
      on_failure: download_failure
    )
  end
```

I get this error on ANY url  I'm fetching...

```
/Users/xxxxx/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/feedjira-1.3.0/lib/feedjira/feed.rb:237: warning: uncaught exception from callback
```

the entire source code for this tutorial app is located here....

https://github.com/thefonso/feedjira-with-rails

And here is some rails console output (I'm using pry)...

```
[2] pry(main)> jiratest = Blog.all
  Blog Load (0.6ms)  SELECT "blogs".* FROM "blogs"
=> [#<Blog id: 1, url: "news.google.com", created_at: "2015-02-24 05:18:50", updated_at: "2015-02-24 05:18:50">,
 #<Blog id: 2, url: "https://news.google.com", created_at: "2015-02-24 05:20:02", updated_at: "2015-02-24 05:20:02">,
 #<Blog id: 4, url: "http://feedjira.com/blog/feed.xml", created_at: "2015-02-24 05:24:20", updated_at: "2015-02-24 05:24:20">,
 #<Blog id: 5, url: "https://news.yahoo.com/rss/topstories", created_at: "2015-02-24 06:26:19", updated_at: "2015-02-24 06:26:19">,
 #<Blog id: 6, url: "http://pandodaily.com.feedsportal.com/c/35141/f/65...", created_at: "2015-02-25 03:49:22", updated_at: "2015-02-25 03:49:22">,
 #<Blog id: 10, url: "http://rss.csmonitor.com/feeds/world", created_at: "2015-03-07 23:44:55", updated_at: "2015-03-07 23:44:55">]
[3] pry(main)> jiratest.update_all_blogs
  Blog Load (0.4ms)  SELECT "blogs".* FROM "blogs"
  Entry Load (0.7ms)  SELECT  "entries".* FROM "entries"  WHERE "entries"."blog_id" = $1  ORDER BY "entries"."published" DESC LIMIT 1  [["blog_id", 1]]
/Users/xxxxx/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/feedjira-1.3.0/lib/feedjira/feed.rb:237: warning: uncaught exception from callback
  Entry Load (0.5ms)  SELECT  "entries".* FROM "entries"  WHERE "entries"."blog_id" = $1  ORDER BY "entries"."published" DESC LIMIT 1  [["blog_id", 2]]
/Users/xxxxx/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/feedjira-1.3.0/lib/feedjira/feed.rb:237: warning: uncaught exception from callback
  Entry Load (0.5ms)  SELECT  "entries".* FROM "entries"  WHERE "entries"."blog_id" = $1  ORDER BY "entries"."published" DESC LIMIT 1  [["blog_id", 4]]
/Users/xxxxx/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/feedjira-1.3.0/lib/feedjira/feed.rb:237: warning: uncaught exception from callback
  Entry Load (1.0ms)  SELECT  "entries".* FROM "entries"  WHERE "entries"."blog_id" = $1  ORDER BY "entries"."published" DESC LIMIT 1  [["blog_id", 5]]
/Users/xxxxx/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/feedjira-1.3.0/lib/feedjira/feed.rb:237: warning: uncaught exception from callback
  Entry Load (0.9ms)  SELECT  "entries".* FROM "entries"  WHERE "entries"."blog_id" = $1  ORDER BY "entries"."published" DESC LIMIT 1  [["blog_id", 6]]
/Users/xxxxx/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/feedjira-1.3.0/lib/feedjira/feed.rb:237: warning: uncaught exception from callback
  Entry Load (0.4ms)  SELECT  "entries".* FROM "entries"  WHERE "entries"."blog_id" = $1  ORDER BY "entries"."published" DESC LIMIT 1  [["blog_id", 10]]
/Users/xxxxx/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/feedjira-1.3.0/lib/feedjira/feed.rb:237: warning: uncaught exception from callback
=> [#<Blog id: 1, url: "news.google.com", created_at: "2015-02-24 05:18:50", updated_at: "2015-02-24 05:18:50">,
 #<Blog id: 2, url: "https://news.google.com", created_at: "2015-02-24 05:20:02", updated_at: "2015-02-24 05:20:02">,
 #<Blog id: 4, url: "http://feedjira.com/blog/feed.xml", created_at: "2015-02-24 05:24:20", updated_at: "2015-02-24 05:24:20">,
 #<Blog id: 5, url: "https://news.yahoo.com/rss/topstories", created_at: "2015-02-24 06:26:19", updated_at: "2015-02-24 06:26:19">,
 #<Blog id: 6, url: "http://pandodaily.com.feedsportal.com/c/35141/f/65...", created_at: "2015-02-25 03:49:22", updated_at: "2015-02-25 03:49:22">,
 #<Blog id: 10, url: "http://rss.csmonitor.com/feeds/world", created_at: "2015-03-07 23:44:55", updated_at: "2015-03-07 23:44:55">]
```

**Comment by amatriain** [comment]
I don't think it's related to Feedjira. Rather I think that sample project has a bug.

Look at https://github.com/thefonso/feedjira-with-rails/blob/master/app/models/blog.rb, for every blog that is updated Feedjira is invoked like this:

```
Feedjira::Feed.fetch_and_parse(
  url,
  on_success: download_success_date,
  on_failure: download_failure
)
```

The ```on_success``` callback is executed after successfully fetching and parsing the feed. The callback is defined as a lambda like this:

```
download_success_date = lambda { |url, feed|
  next unless !@newest_entry || entry.published > @newest_entry.published
  add_new_entry_from_feed(entry)
}
```

As you can see it's using a variable named ```entry```, but there is no variable with that name in the scope. I think trying to invoke the ```.published``` method on that undefined variable raises the error you're seeing: ```warning: uncaught exception from callback```.

It looks like that callback is supposed to perform some action for each new entry that has been fetched, but not for older entries. The way it's written doesn't make much sense, I'm guessing it should be something like this instead:

```
download_success_date = lambda { |url, feed|
  feed.entries.each do |entry|
    next unless !@newest_entry || entry.published > @newest_entry.published
    add_new_entry_from_feed(entry)
  end
}
```

At least when I write the lambda like this it no longer raises an error. I haven't looked much at the code so I cannot vouch if it now works as it's supposed to.

TL;DR it is not a Feedjira error. It is an error in the lambda you're passing to the on_success callback.

**Comment by thefonso** [comment]
yep binding.pry would seem to agree. thank you.
